programs/host/frontend.c

```diff
@@ -0,0 +1,957 @@
+                offset = item_uint32( inputs[0] );
+                whence = item_uint32( inputs[1] );
+                file_id = item_uint32( inputs[2] );
+                file = (FILE *)get_file_from_id( file_id );
+
+                if ( file )
+                {
+                    ret = 0;
+
+                    /* map whence with std lib */
+                    switch ( whence )
+                    {
+                        case MBEDTLS_SERIALIZE_FSEEK_SET:
+                            whence = SEEK_SET;
+                            break;
+                        case MBEDTLS_SERIALIZE_FSEEK_CUR:
+                            whence = SEEK_CUR;
+                            break;
+                        case MBEDTLS_SERIALIZE_FSEEK_END:
+                            whence = SEEK_END;
+                            break;
+                        default:
+                            ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+                            break;
+                    }
+
+                    if ( ret == 0 )
+                        ret = fseek( file, offset, whence );
+                }
+            }
+            break;
+        case MBEDTLS_SERIALIZE_FUNCTION_FTELL:
+            {
+                FILE * file = NULL;
+                int32_t file_id;
+
+                ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+                file_id = item_uint32( inputs[0] );
+                file = (FILE *)get_file_from_id( file_id );
+
+                if ( file )
+                {
+                    ret = ftell( file );
+
+                    if ( ret >= 0 )
+                    {
+                        ALLOC_OUTPUT( 0, sizeof( int32_t ) );
+                        set_item_uint32( outputs[0], ret );
+                        ret = 0;
+                    }
+                }
+            }
+            break;
+        case MBEDTLS_SERIALIZE_FUNCTION_FERROR:
+            {
+                FILE * file = NULL;
+                int32_t file_id;
+
+                ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+                file_id = item_uint32( inputs[0] );
+                file = (FILE *)get_file_from_id( file_id );
+
+                if ( file )
+                {
+                    ret = ferror( file );
+                }
+            }
+            break;
+        case MBEDTLS_SERIALIZE_FUNCTION_DOPEN:
+            {
+                char * path = NULL;
+                int32_t file_id;
+
+                ret = MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT;
+                path = item_buffer( inputs[0] );
+                DBG( "open dir [%s]", path);
+                file_id = alloc_file_context();
+                DBG( "allocated dir id [%d]", file_id);
+                if ( file_id != -1 )
+                {
+                    DIR * dir = opendir( path );
+                    if ( dir != NULL )
+                    {
+                        files[file_id - 1].file = (void *)dir;
+                        ALLOC_OUTPUT( 0, sizeof (int32_t) );
```

Comment:
`ALLOC_OUTPUT` needs to go before `alloc_file_context` and `opendir`, same reasons as with `fopen`.